Add EIP-5678: Add new utilities extension for ERC-721 (#5678) (#6785)
ethereum/EIPs: EIPS/eip-6785.md

---
eip: 6785
title: ERC-721 Utilities Extension

> description: Easy access to utility information via functions and the NFT's metadata

Reviewer: The description as written doesn't give much more information than the title. If you can cram what a "utility function" is in here I think it would be very helpful.

author: Otniel Nicola (@OT-kthd), Bogdan Popa (@BogdanKTHD)
discussions-to: https://ethereum-magicians.org/t/eip-6785-erc-721-utilities-extension/13568
status: Draft
type: Standards Track
category: ERC
created: 2023-03-27
requires: 165, 721
---

## Abstract

This specification defines standard functions that outline what a utility entails for a specific NFT and how it may be

> used. This specification is an optional extension of [ERC-721](./eip-721.md).

Reviewer: You should make it extremely obvious in the abstract that a "utility" is a property of a token, and that you aren't defining "utility functions for interacting with tokens."

    event UpdateUtility(uint256 indexed tokenId, string utilityUri);

    /// @notice set the new utilityUri - remember the date it was set on
    /// @dev The empty string indicates there is no utility
    /// Throws if `tokenId` is not valid NFT
    /// @param utilityUri  The new utility description of the NFT
    /// 324a28b4
    function setUtilityUri(uint256 tokenId, string utilityUri) external;

    /// @notice Get the utilityUri of an NFT
    /// @dev The empty string for `utilityUri` indicates that there is no utility associated
    /// @param tokenId The NFT to get the user address for
    /// @return The utility uri for this NFT
    /// 438b8582
    function getUtilityUri(uint256 tokenId) external view returns (string memory);

    /// @notice Get the changes made to utilityUri
    /// @param tokenId The NFT to get the user address for
    /// @return The history of changes to `utilityUri` for this NFT
    /// af5a3ae2
    function getUtilityHistory(uint256 tokenId) external view returns (string[] memory);
}
```

All functions defined as view MAY be implemented as pure or view

Function `setUtilityUri` MAY be implemented as public or external

> The event `UpdateUtility` MUST be emitted when the setUtilityUri function is called

Reviewer: Should it also be emitted for changes not caused by `setUtilityUri`? Say in the constructor, or if a contract has some custom batch update function?

becomes unavailable or inaccessible.

As the instructions on how to access a given utility may change over time, there should be a historical record of these
changes for transparency.

For example, if a creator sells an NFT that gives holders a right to a video call with the creator, the metadata for
this utility NFT would read as follows:

```json
{
  "name": "...",
  "description": "...",
  "image": "...",
  "utilities": {
    "type": "Video call",
    "description": "I will enter a private video call with whoever owns the NFT",
    "properties": [
      {
        "sessions": 2
      },
      {
        "duration": 30
      },
      {
        "time_unit": "minutes"
      }
    ],
    "expiry": 1.577e+7,
    "t&c": "https://...."
  }
}
```

In order to get access to the details needed to enter the video call, the owner would access the URI returned by
the `getUtilityUri` method for the NFT that they own. Additionally, access to the details could be conditioned by the
authentication with the wallet that owns the NFT.
The current status of the utility would also be included in the URI (eg: how many sessions are still available, etc.)

> 

Reviewer: Some of this should move to motivation.

---
eip: 6785
title: ERC-721 Utilities Information Extension
description: Provide easy access to information about the `utility` of an NFT via functions and the NFT's metadata
author: Otniel Nicola (@OT-kthd), Bogdan Popa (@BogdanKTHD)
discussions-to: https://ethereum-magicians.org/t/eip-6785-erc-721-utilities-extension/13568
status: Draft
type: Standards Track
category: ERC
created: 2023-03-27
requires: 165, 721
---

## Abstract

This specification defines standard functions and an extension of the metadata schema that outline what a 
token's utility entails and how the utility may be used and/or accessed.
This specification is an optional extension of [ERC-721](./eip-721.md).

## Motivation

> This specification aims to bring clarity what the utility associated with an NFT is and how to access this utility.

Reviewer: ```suggestion
This specification aims to clarify what the utility associated with an NFT is and how to access this utility.
```

Relying on third-party platforms to obtain information regarding the utility of the NFT that one owns can lead to scams,
phishing or other forms of fraud.

> Currently, utilities that are offered with NFTs, are not captured on-chain. We want the utility of an NFT to be part of

Reviewer: ```suggestion
Currently, utilities that are offered with NFTs are not captured on-chain. We want the utility of an NFT to be part of
```

> the metadata of an NFT, and thus immutable. The metadata information would include: a) type of utility, b) description

Reviewer: I'd be careful about claiming immutability here, since you're using URIs. The content located by the URI can be changed, unless it's a `data:`, `ipfs:`, or similarly immutable scheme.

    /// @dev The empty string indicates there is no utility
    /// Throws if `tokenId` is not valid NFT
    /// @param utilityUri  The new utility description of the NFT
    /// 4a048176
    function setUtilityUri(uint256 tokenId, string utilityUri) external;

    /// @notice Get the utilityUri of an NFT
    /// @dev The empty string for `utilityUri` indicates that there is no utility associated
    /// @param tokenId The NFT to get the user address for
    /// @return The utility uri for this NFT
    /// 5e470cbc
    function utilityUriOf(uint256 tokenId) external view returns (string memory);

    /// @notice Get the changes made to utilityUri
    /// @param tokenId The NFT to get the user address for
    /// @return The history of changes to `utilityUri` for this NFT
    /// f96090b9
    function utilityHistoryOf(uint256 tokenId) external view returns (string[] memory);
}
```

All functions defined as view MAY be implemented as pure or view

Function `setUtilityUri` MAY be implemented as public or external

The event `UpdateUtility` MUST be emitted when the `setUtilityUri` function is called or any other time that the utility
of the token is changed, like in batch updates. 

> The `supportsInterface` method MUST return true when called with `ed231d73`

Reviewer: This interface id doesn't match the comment on line 45. Is that intentional?

Author: the comment was unchanged from the previous value, before modifying the interface

The original metadata SHOULD conform to the “ERC-6785 Metadata with utilities JSON Schema” which is a compatible

> extension of the “ERC-721 Metadata JSON Schema” defined in `ERC-721`.

Reviewer: ```suggestion
extension of the “ERC-721 Metadata JSON Schema” defined in ERC-721.
```

        },
        "properties": {
          "type": "array",
          "description": "An array of possible properties describing the utility, defined as key-value pairs",
          "items": {
            "type": "object"
          }
        },
        "expiry": {
          "type": "number",
          "description": "The period of time for the validity of the utility, since the minting of the NFT. Expressed in seconds"
        },
        "t&c": {
          "type": "string",
          "description": ""
        }
      }
    }
  }
}
```

## Rationale

Instructions on how to access the utility should be available to the owner of an NFT, as it reduces the dependency on
third-party platforms and the possibility for fraud because the `utilityUri` containing the instructions can only come
from the entity that provides or facilitates the utility. Thus, the ability to set the `utilityUri` should be restricted
to the creator who's offering the utility. Since the `utilityUri` contains information that could be restricted to some
level and is dependent on a third party platform, the creator needs the ability to modify it in the event the platform

> becomes unavailable or inaccessible.

Reviewer: Some of this paragraph seems to be adding requirements (sentences containing "should", "may", etc.) Requirements on implementations should go in the specification section.

The rationale section should explain why you made certain technical decisions within the EIP itself.

---
eip: 6785
title: ERC-721 Utilities Information Extension
description: Provide easy access to information about the `utility` of an NFT via functions and the NFT's metadata
author: Otniel Nicola (@OT-kthd), Bogdan Popa (@BogdanKTHD)
discussions-to: https://ethereum-magicians.org/t/eip-6785-erc-721-utilities-extension/13568
status: Draft
type: Standards Track
category: ERC
created: 2023-03-27
requires: 165, 721
---

## Abstract

> This specification defines standard functions and an extension of the metadata schema that outline what a 

Reviewer: ```suggestion
This specification defines standard functions and an extension of the metadata schema that outlines what a 
```
